### PR TITLE
[python-frontend] Handling module aliases

### PR DESCRIPTION
**Pull request by brcfarias**
This PR enables type inference and verification for functions imported and invoked through aliases.

- Add a module manager to extract function return type from imported modules.
- Support function calls from aliased modules

**Example**
**_other.py_**
```python
def sum(a:int, b:int) : int ->
  return a+b
```

_**main.py:**_
```python
import other as ot
x = 1
y = 2
z = ot.sum(x,y) # imported function with alias
```

### CODE REVIEW

File: regression/python/import-as/test.desc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+CORE
+main.py
+
+^VERIFICATION SUCCESSFUL$
```
Comment (lucasccordeiro) [inline regression/python/import-as/test.desc:4]: Can you also add a test case that provides `VERIFICATION FAILED`? For example, you could check for overflow.

File: src/python-frontend/json_utils.h [hunk 2]
```diff
@@ -94,6 +96,39 @@ JsonType &find_function(JsonType &json, const std::string &func_name)
...
+    }
+  }
+  throw std::runtime_error("Function " + func_name + " not found\n");
+}
+
+template <typename JsonType>
+const std::string
+get_object_alias(const JsonType &ast, const std::string &obj_name)
+{
+  for (auto &node : ast["body"])
```
Comment (lucasccordeiro) [inline src/python-frontend/json_utils.h:119]: This loop iterates through all nodes and names, which can become expensive for large ASTs. 

Could we consider stopping early once the target is found or preprocessing the AST into a more query-efficient structure?

File: src/python-frontend/json_utils.h [hunk 3]
```diff
@@ -94,6 +96,39 @@ JsonType &find_function(JsonType &json, const std::string &func_name)
   throw std::runtime_error("Function " + func_name + " not found\n");
 }
 
+template <typename JsonType>
+JsonType &find_imported_function(JsonType &ast, const std::string &func_name)
```
Comment (lucasccordeiro) [inline src/python-frontend/json_utils.h]: Shall we consider `const JsonType &ast`?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]

File: src/python-frontend/module.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  void add_submodule(const std::shared_ptr<module> mod)
+  {
+    submodules_.insert(mod);
+  }
+
+  function get_function(const std::string &func_name) const
+  {
+    auto it = std::find_if(
```
Comment (lucasccordeiro) [inline src/python-frontend/module.h:60]: Could we use a `std::unordered_map` or `std::map` instead of `std::vector`? This would enable O(1) (unordered_map) or O(log n) (map) lookups.

File: src/python-frontend/module.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    return {};
+  }
+
+  const std::vector<function> &functions() const
+  {
+    return functions_;
+  }
+
+  const std::unordered_set<std::shared_ptr<module>> &submodules() const
```
Comment (lucasccordeiro) [inline src/python-frontend/module.h]: Shall we consider std::unique_ptr? If shared ownership is required, consider documenting why shared ownership is necessary here.

File: src/python-frontend/python_annotation.h [hunk 6]
```diff
@@ -330,28 +348,76 @@ class python_annotation
...
     std::string type("");
 
-    const std::string &obj = get_object_name(call["func"]);
+    const std::string &obj = get_object_name(call["func"], std::string());
 
-    python_module pm(obj);
-    if (pm.is_standard_module())
-      return pm.get_function_return(call["func"]["attr"]);
+    // Get type from imported module
+    if (mm_)
```
Comment (lucasccordeiro) [inline src/python-frontend/python_annotation.h]: Shall we use a more suggestive name for this `_mm`?

File: src/python-frontend/json_utils.h (hunk 2 above)
Comment (brcfarias) [reply src/python-frontend/json_utils.h:119]: Thanks for the suggestion. The function stops early with a return once the target alias is found.

File: src/python-frontend/module_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+      if (!current_module && !module_name.empty())
+      {
+        current_module = std::make_shared<module>(module_name);
+        modules_.insert(current_module);
+      }
+    }
+    else
+    {
+      // If there is a parent module, create or get it as a submodule
+      auto existing_submodule = std::find_if(
```
Comment (lucasccordeiro) [inline src/python-frontend/module_manager.cpp:143]: Shall we use an associative container (`std::unordered_map`) for faster lookups?

File: src/python-frontend/module_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+void module_manager::load_directory(
+  const fs::path &current_path,
+  std::shared_ptr<module> parent_module)
+{
+  for (const auto &entry : fs::directory_iterator(current_path))
+  {
+    if (entry.is_regular_file() && entry.path().extension() == ".json")
+    {
+      // Create a module for the JSON file
+      auto submodule = create_module(entry.path());
```
Comment (lucasccordeiro) [inline src/python-frontend/module_manager.cpp:79]: Shall we wrap the JSON parsing code in a try-catch block to handle exceptions?

**Review by lucasccordeiro — APPROVED** [review]
Thanks for addressing the comments, @brcfarias.

I have reviewed the code again. I added just a few suggestions to consider for improvement.

Apart from that, LGTM.

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @brcfarias.
